ts/features/idpay/unsubscription/xstate/services.ts

```diff
@@ -0,0 +1,75 @@
+/* eslint-disable no-underscore-dangle */
+import * as E from "fp-ts/lib/Either";
+import { pipe } from "fp-ts/lib/function";
+import { PreferredLanguage } from "../../../../../definitions/backend/PreferredLanguage";
+import { InitiativeDTO } from "../../../../../definitions/idpay/InitiativeDTO";
+import { IDPayClient } from "../../common/api/client";
+import { Context } from "./context";
+
+export type Services = {
+  getInitiativeInfo: {
+    data: InitiativeDTO;
+  };
+  unsubscribeFromInitiative: {
+    data: undefined;
+  };
+};
+
+const createServicesImplementation = (
+  client: IDPayClient,
+  token: string,
+  language: PreferredLanguage
+) => {
+  const getInitiativeInfo = async (context: Context) => {
+    const dataResponse = await client.getWalletDetail({
+      bearerAuth: token,
+      "Accept-Language": language,
+      initiativeId: context.initiativeId
+    });
+
+    const data: Promise<InitiativeDTO> = pipe(
+      dataResponse,
+      E.fold(
+        _ => Promise.reject(undefined),
+        response => {
+          if (response.status !== 200) {
+            return Promise.reject(undefined);
+          }
+          return Promise.resolve(response.value);
+        }
+      )
+    );
+
+    return data;
+  };
```

Comment:
Maybe this could be refactor using TE from fp-ts and improve code readability.

Reply:
Definitely. I'll try to refactor the services making them more functional.

Reply:
@hevelius I tried to find a proper services implementation using a functional approach (using TaskEither) but I found it a bit too complicated and, in the end, it wasn't worth it..

For the XState machine to work properly, services must return a `Promise<T>`. This is easily done with a `TaskEither`, but the client also returns an `Either` (due to its internal validation), which needs to be mapped/folded.
This results in a lot of code to replace very few lines. Maybe it's just me not having enough experience with `fp-ts`.
Feel free to suggest a solution using `TaskEither`.

By the way, I added a try-catch to the services to correctly catch any networking problem caused by the API calls. This should make them more robust.